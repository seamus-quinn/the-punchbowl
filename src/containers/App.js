import React, { Component } from 'react';
import * as firebase from 'firebase';
import { connect } from 'react-redux';

import './App.css';
import * as apiCalls from '../apiCalls';
import { populateArticles } from '../actions';

class App extends Component {
  constructor(props) {
    super(props);

    this.state = {
      domains: ['npr.org', 'nytimes.com', 'nbcnews.com', 'theatlantic.com', 'ap.org', 'c-span.org', 'foxnews.com', 'wsj.com', 'cato.org', 'breitbart.com'],
      matches: [],
    }
  }

  async componentDidMount() {
    // console.log(firebase.database())
    // // firebase.database().ref('mockData/').set({
    // //   seamus: 'garbage',
    // //   garbage: 'seamus'
    // // })
    // await this.fetchAllArticles()
  }

  fetchAllArticles = async () => {
    const { domains } = this.state
    const articles = []
    domains.map(async domain => {
      for (let i = 1; i < 2; i++) {
        const articlesToStore = await apiCalls.fetchArticles(domain, i)
        articles.push(...articlesToStore.articles)
      }
    })
    this.props.populateArticles(articles)
  }

  render() {
    return (
      <div className="App">
        <header className="App-header">
          <h1 className="App-title">Welcome to React</h1>
        </header>
        <p className="App-intro">
          To get started, edit <code>src/App.js</code> and save to reload.
        </p>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  populateArticles: (articles) => dispatch(populateArticles(articles))
})

const config = {
  apiKey: "AIzaSyDiBMqPsE5G7LQrVzd0SNdf_64Zi9zRgfI",
  authDomain: "the-punchbowl.firebaseapp.com",
  databaseURL: "https://the-punchbowl.firebaseio.com",
  projectId: "the-punchbowl",
  storageBucket: "the-punchbowl.appspot.com",
  messagingSenderId: "126203247142"
};

if (!firebase.apps.length) {
  firebase.initializeApp(config)
}

export {
  App,
  mapDispatchToProps
}

export default connect(null, mapDispatchToProps)(App)
