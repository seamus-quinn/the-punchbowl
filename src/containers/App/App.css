.App {
  text-align: center;
  background: rgb(84,95,164);
  background: linear-gradient(214deg, rgba(84,95,164,1) 0%, rgba(114,173,207,1) 53%);
  height: 100%;
}

.header {
  height: 100px;
  color: #fff;
  display: flex;
  justify-content: flex-end;
  align-items: center;
}

.app-title {
  font-size: 40px;
  margin: 0 3%;
  font-family: 'Poiret One', cursive;
  letter-spacing: 2px;
}

.border {
  height: 1px;
  background: rgb(103,117,176);
  background: linear-gradient(214deg, rgba(103,117,176,1) 0%, rgb(153, 195, 231) 100%);
  margin: 0 3%;
}


