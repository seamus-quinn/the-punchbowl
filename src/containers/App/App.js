import React, { Component } from 'react';
import * as firebase from 'firebase';
import { connect } from 'react-redux';

import './App.css';
import * as apiCalls from '../../apiCalls';
import * as mockData from '../../mockData';
import { populateArticles } from '../../actions';

import Search from '../Search/Search'
import MatchContainer from '../MatchContainer/MatchContainer'

class App extends Component {
  constructor(props) {
    super(props);

    this.state = {
      domains: ['npr.org', 'nytimes.com', 'nbcnews.com', 'theatlantic.com', 'ap.org', 'c-span.org', 'foxnews.com', 'wsj.com', 'cato.org', 'breitbart.com'],
    }
  }

  fetchAllArticles = async () => {
    const { domains } = this.state
    const allArticles = domains.map(async domain => {
      const domainArticles = []
      for (let i = 1; i < 10; i++) {
        const fetchedArticles = await apiCalls.fetchArticles(domain, i)
        domainArticles.push(...fetchedArticles.articles)
      }
      return domainArticles;
    })
    return await Promise.all(allArticles)
  }

  async componentDidMount() {
    const ref = firebase.database().ref('/')
    ref.on('value', this.getData);
  }

  flattenArrays = (arr) => {
    return arr.reduce( (acc, arr) => {
      return acc.concat(Array.isArray(arr) ? this.flattenArrays(arr) : arr);
    }, []);
  }

  getData = async (data) => {
    const currentTime = Date.now();
    const { articles, timeStamp } = data.val();
    console.log(currentTime - timeStamp)
    if (currentTime - timeStamp >= 43200000) {
      const nestedArticles = await this.fetchAllArticles();
      const articles = this.flattenArrays(nestedArticles)
      this.props.populateArticles(articles)
      firebase.database().ref('/').set({
        timeStamp: currentTime,
        articles
      })
      console.log('ive done it!')
    } else {
      this.props.populateArticles(articles)
    }
  }


  render() {
    return (
      <div className="App">
        <header className='header'>
          <h1 className='app-title'>the punchBowl</h1>
        </header>
        <div className='border'></div>
        <Search />
        <MatchContainer />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  populateArticles: (articles) => dispatch(populateArticles(articles))
})

const config = {
  apiKey: "AIzaSyDiBMqPsE5G7LQrVzd0SNdf_64Zi9zRgfI",
  authDomain: "the-punchbowl.firebaseapp.com",
  databaseURL: "https://the-punchbowl.firebaseio.com",
  projectId: "the-punchbowl",
  storageBucket: "the-punchbowl.appspot.com",
  messagingSenderId: "126203247142"
};

if (!firebase.apps.length) {
  firebase.initializeApp(config)
}

export {
  App,
  mapDispatchToProps
}

export default connect(null, mapDispatchToProps)(App)
