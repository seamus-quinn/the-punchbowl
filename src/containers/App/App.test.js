import React from 'react';
import { shallow } from 'enzyme';

// import from './App'
import { App, mapDispatchToProps } from './App'

describe('App', () => {
  let wrapper
  let mockProps

  beforeEach(() => {
    mockProps = {
      populateArticles: jest.fn()
    }
    wrapper = shallow(<App {...mockProps}/>)
  })

  it('matches the snapshot', () => {

    expect(wrapper).toMatchSnapshot();
  })

  describe('componentDidMount', () => {

  })

  describe('fetchAllArticles', () => {

  })

  describe('checkFirebase', () => {
    
    let mockData;

    beforeEach(() => {
      mockData = {
        articles: [
          {title: 'Roseanne pops an ambien and calls Trump a peach'},
          {title: 'J.R. Smith forgets to put on his shoes before Game 2'}
        ],
        timeStamp: 1
      }
    })

    it('calls populateArticles with the correct arguments if the elapsed time is less than 12 hours', () => {
      const wrapperInst = wrapper.instance();

      Date.now = jest.fn().mockImplementation(() => 4)

      mockData.val = jest.fn().mockImplementation(() => mockData)

      wrapperInst.checkFirebase(mockData)

      expect(wrapperInst.props.populateArticles).toHaveBeenCalledWith(mockData.articles)
    })

    it('calls fetchAllArticles if the elapsed time is greater than or equal to 12 hours', () => {

    })

    it('flattens the arrays received from fetchAllArticles if the the elapsed time is greater than or equal to 12 hours', () => {

    })

    it('calls populateArticles with the correct arguments if the elapsed time is greater than or equal to 12 hours', () => {

    })

    it('sends updated information to firebase database if the elapsed time is greater than or equal to 12 hours', () => {

    })

  })

  describe('mapDispatchToProps', () => {

    it('returns an object with a populateArticles function', () => {
      const dispatch = jest.fn();
      const result = mapDispatchToProps(dispatch);

      expect(typeof result.populateArticles).toEqual('function')

    })

    it('calls dispatch with the correct arguments', () => {
      const dispatch = jest.fn();
      const result = mapDispatchToProps(dispatch);
      const mockAction = {
        type: 'POPULATE_ARTICLES',
        articles: ['wow', 'grandma', 'you', 'so', 'strong']
      }

      result.populateArticles(mockAction.articles);

      expect(dispatch).toHaveBeenCalledWith(mockAction)
    })
  })

})