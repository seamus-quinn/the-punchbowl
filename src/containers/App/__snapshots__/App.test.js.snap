// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <header
          className="header"
>
          <h1
                    className="app-title"
          >
                    the punchBowl
          </h1>
</header>,
        <div
          className="border"
/>,
        <Connect(Search) />,
        <Connect(MatchContainer) />,
      ],
      "className": "App",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <h1
            className="app-title"
>
            the punchBowl
</h1>,
          "className": "header",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "the punchBowl",
            "className": "app-title",
          },
          "ref": null,
          "rendered": "the punchBowl",
          "type": "h1",
        },
        "type": "header",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "border",
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <header
            className="header"
>
            <h1
                        className="app-title"
            >
                        the punchBowl
            </h1>
</header>,
          <div
            className="border"
/>,
          <Connect(Search) />,
          <Connect(MatchContainer) />,
        ],
        "className": "App",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <h1
              className="app-title"
>
              the punchBowl
</h1>,
            "className": "header",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "the punchBowl",
              "className": "app-title",
            },
            "ref": null,
            "rendered": "the punchBowl",
            "type": "h1",
          },
          "type": "header",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "border",
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
